amaranth/vendor/lattice_nexus.py

```diff
@@ -0,0 +1,675 @@
+        assert False
+
+    @property
+    def default_clk_constraint(self):
+        if self.default_clk == "OSC":
+            return Clock(150e6 / self.osc_div) # lscc/radiant/2022.1/docs/webhelp/eng/index.htm#page/Reference%20Guides/FPGA%20Libraries%20Reference%20Guide/Alphanumeric_Primitives_List.11.155.htm        
+        # For HFCLKOUT (created clock): ADJUSTED_PERIOD=6666.66667ps * HF_CLK_DIV
+        return super().default_clk_constraint
+
+    def create_missing_domain(self, name):
+        # Lattice ECP5 devices have two global set/reset signals: PUR, which is driven at startup
+        # by the configuration logic and unconditionally resets every storage element, and GSR,
+        # which is driven by user logic and each storage element may be configured as affected or
+        # unaffected by GSR. PUR is purely asynchronous, so even though it is a low-skew global
+        # network, its deassertion may violate a setup/hold constraint with relation to a user
+        # clock. To avoid this, a GSR/SGSR instance should be driven synchronized to user clock.
+        if name == "sync" and self.default_clk is not None:
+            m = Module()
+            if self.default_clk == "OSC":
+                if not hasattr(self, "osc_div"):
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 "
+                                     "and 128")
+                if not isinstance(self.osc_div, int) or self.osc_div < 2 or self.osc_div > 128:
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 "
+                                     "and 128, not {!r}"
+                                     .format(self.osc_div))
+                clk_i = Signal()
+                m.submodules += Instance("OSC", p_HF_CLK_DIV=self.osc_div, o_HFCLKOUT=clk_i)
+            else:
+                clk_i = self.request(self.default_clk).i
+            if self.default_rst is not None:
+                rst_i = self.request(self.default_rst).i
+            else:
+                rst_i = Const(0)
+
+            gsr0 = Signal()
+            gsr1 = Signal()
+            # There is no end-of-startup signal on ECP5, but PUR is released after IOB enable, so
+            # a simple reset synchronizer (with PUR as the asynchronous reset) does the job.
+            # LIFCL UPDATE: GSR() has a parameter (SYNCMODE) to perform synchronization of the reset to the user clock
+            m.submodules += [
+                #Instance("FD1S3AX", p_GSR="DISABLED", i_CK=clk_i, i_D=~rst_i, o_Q=gsr0),
+                #Instance("FD1S3AX", p_GSR="DISABLED", i_CK=clk_i, i_D=gsr0,   o_Q=gsr1),
+                ## Although we already synchronize the reset input to user clock, SGSR has dedicated
+                ## clock routing to the center of the FPGA; use that just in case it turns out to be
+                ## more reliable. (None of this is documented.)
+                #Instance("SGSR", i_CLK=clk_i, i_GSR=gsr1),
```

Comment:
Looks good--I think you can delete the pre-LIFCL code and only leave the current GSR configuration

Reply:
ok

Reply:
I don't think nextpnr-nexus ever gained GSR support, fyi

Reply:
I see. I was just checking the primitives reference guide for the crosslink (`lscc/radiant/2022.1/docs/webhelp/eng/Reference Guides/`). You're right tho, pressing rst on the board does nothing.